pkg/scheduler/framework/frameworkutils.go

```diff
@@ -0,0 +1,372 @@
+//   - active bindings, i.e., bindings that are associated with a normally operating cluster and
+//     have been cleared for processing by the dispatcher; and
+//   - creating bindings, i.e., bindings that have been associated with a normally operating cluster,
+//     but have not yet been cleared for processing by the dispatcher; and
+//   - dangling bindings, i.e., bindings that are associated with a cluster that is no longer in
+//     a normally operating state (the cluster has left the fleet, or is in the state of leaving),
+//     yet has not been marked as deleting by the scheduler; and
+//   - obsolete bindings, i.e., bindings that are no longer associated with the latest scheduling
+//     policy; and
+//   - deleted bindings, i.e, bindings that are marked for deletion in the API server, but have not
+//     yet been marked as deleting by the scheduler.
+func classifyBindings(policy *fleetv1beta1.ClusterPolicySnapshot, bindings []fleetv1beta1.ClusterResourceBinding, clusters []fleetv1beta1.MemberCluster) (active, creating, obsolete, dangling, deleted []*fleetv1beta1.ClusterResourceBinding) {
+	// Pre-allocate arrays.
+	active = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+	creating = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+	obsolete = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+	dangling = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+	deleted = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+
+	// Build a map for clusters for quick loopup.
+	clusterMap := make(map[string]fleetv1beta1.MemberCluster)
+	for _, cluster := range clusters {
+		clusterMap[cluster.Name] = cluster
+	}
+
+	for idx := range bindings {
+		binding := bindings[idx]
+		targetCluster, isTargetClusterPresent := clusterMap[binding.Spec.TargetCluster]
+
+		switch {
+		case binding.DeletionTimestamp != nil:
+			// Check if the binding has been deleted, but still has the scheduler cleanup finalizer in presence.
+			if controllerutil.ContainsFinalizer(&binding, fleetv1beta1.SchedulerCleanupFinalizer) {
+				deleted = append(deleted, &binding)
+			}
+			// Ignore any binding that has been deleted, and has no scheduler cleanup finalizer in presence.
+		case binding.Spec.State == fleetv1beta1.BindingStateDeleting:
+			// Ignore any binding that is of the deleting state.
+		case !isTargetClusterPresent || targetCluster.Spec.State == fleetv1beta1.ClusterStateLeave:
+			// Check if the binding is now obsolete, i.e., it is associated with a cluster that is no longer
+			// in normal operations, but is still of an active or creating state.
+			//
+			// Note that this check is solely for the purpose of detecting a situation where bindings are stranded
+			// on a leaving/left cluster; it does not perform any binding association eligibility check for the cluster.
+			dangling = append(dangling, &binding)
+		case binding.Spec.PolicySnapshotName != policy.Name:
```

Comment:
it seems that this is bound to happen for most of the bindings when a new policy is to be scheduled. None of the existing bind would have the new policy name, right?

Reply:
The inconsistency could happen if the scheduler fails when creating the new bindings.